### PR TITLE
shell: fix possible hanging reason when history feature is used

### LINKED ISSUE #10413: Shell: trying to browse history freezes shell on disco_l475_iot1
can be reproduced in samples/subsys/shell/shell_module when built for the disco_l475_iot1, run one command, for example "kernel threads", then try to re-run by pressing up arrow, the shell freezes.

Works fine on nrf boards.

### PR DESCRIPTION
**Pull request by jakub-uC**
Fixing shell hanging reason.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

### CODE REVIEW

**Review by nordic-krch — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@jukkar can you test this please?

**Comment by jarz-nordic** [comment]
Until shell rx will be interrupt driven, just for testing purpose you can try to increase shell's thread priority. This should help to not loose input charcaters.

**Comment by nashif** [comment]
does not fix it for me on disco_l475_iot1

**Comment by jarz-nordic** [comment]
@nashif : What does it mean? Shell is still hanging or only history does not work?

**Comment by nashif** [comment]
@jarz-nordic it hangs

**Comment by jarz-nordic** [comment]
This PR makes no sense. In fact I'hve been changing FMS state so stucking in default was not possible.

I have kind request to people who are observing haning issue. May you please debug and put a trap in shell function `shell_state_collect` ? I do not have this board so i am not able to do it myself.

Please let me know if, while handing:
1. this function is executed when you try to type something?
2. does this function receive anything from uart driver: 

```
(void)shell->iface->api->read(shell->iface, &data,
					      sizeof(data), &count);
if (count == 0) {
        return;
} 
```
?

**Comment by zerotri** [comment]
@jarz-nordic I am seeing the code get to the function mentioned.

Shortly after pressing the up key, `shell->iface->api->read` returns with `data` to `27`, `'\033'` and a `count` of `1`. Flow then continues into `receive_state_change` with `state=SHELL_RECEIVE_ESC`.

The next time `shell_state_collect` loops around, `shell->iface->api->read` returns with no change to `data` and a count of `0`. After attempting to write out some data to the shell, flow continues to where `shell_thread` does the following and then never seems to return on further keypress:

`err = k_poll(shell->ctx->events, SHELL_SIGNALS, K_FOREVER);`

It seems to me that the whole excape sequence may not be getting sent through?

I've also tested sending a capital A in and that doesn't seem to resume the shell from this state. It seems that the system never returns to the thread after this.

**Comment by jarz-nordic** [comment]
@zerotri : Thank you for help. Capital `A` won't help, it was false trail :(.

Your observation is ok, debugging will be more tricky because as for now shell is polling UART. As a result when you put a trap after first received character, program will stop on the trap and further characters of the escape sequence will be lost.

Please remove all traps and check what is last state that shell executed?
Please first put one trap in state: `SHELL_RECEIVE_ESC`. Next please remove it and put other trap in state `SHELL_RECEIVE_ESC_SEQ`.

And finall tests, please:
1. remove all traps and run shell.
2. execute some command.
3. hang shell with `up` arrow.
4. put trap just after:
```
if (count == 0) {
        return;
} 
```
and check if shell receives any characters?

Once more thank you for help.

**Comment by zerotri** [comment]
I ran the debugger, breaking on those switch states. Neither ever gets entered.

I decided to step through starting with the state `SHELL_RECEIVE_DEFAULT`.

1. After pressing up, `shell->iface->api->read` sets `data` to `27`, `count` to `1`.
1. This puts it into the state `SHELL_VT100_ASCII_ESC`, where it calls `receive_state_change`.
1. `shell->ctx->receive_state` transitions from `SHELL_RECEIVE_DEFAULT` -> `SHELL_RECEIVE_ESC`
1. Next loop through, `shell->iface->api->read` sets `count` to `0`.
1. Function returns because `count == 0`

This just causes `shell_thread` to hang on`err = k_poll(shell->ctx->events, SHELL_SIGNALS, K_FOREVER)` again.

**Comment by jarz-nordic** [comment]
@zerotri @nashif @carlescufi : I think I found real root cause and I've applied a solution. This PR is a temporary solution until we will have ring buffer in the shell and rx with interrupts.

**Comment by jarz-nordic** [comment]
May you please check if shell is still hanging?

File: subsys/shell/shell_uart.c [hunk 1]
```diff
@@ -73,14 +73,17 @@ static int read(const struct shell_transport *transport,
 		void *data, size_t length, size_t *cnt)
 {
 	struct shell_uart *sh_uart = (struct shell_uart *)transport->ctx;
+	u32_t key;
 
+	key = irq_lock();
 	if (sh_uart->rx_cnt) {
 		memcpy(data, sh_uart->rx, 1);
```
Comment (carlescufi) [inline subsys/shell/shell_uart.c]: Not a part of this PR, but why are you using memcpy() to copy a single byte?

**Comment by jarz-nordic** [comment]
ping

**Comment by nashif** [comment]
yes, it is still an issue, backspace immediately causes the freeze

**Comment by zerotri** [comment]
Hi @jarz-nordic, 
Yes, looks like it's still an issue. Not really seeing any change in the debugger's flow.
Not having any of the issues with backspace as mentioned above though, just the arrow key for history.

**Comment by jarz-nordic** [comment]
I shall have this board in 2 weeks. I will back to this issue shortly after IoT summit in Edinburgh.

**Comment by nashif** [comment]
any updates?

**Comment by carlescufi** [comment]
> any updates?

Have you seen: https://github.com/zephyrproject-rtos/zephyr/issues/10413#issuecomment-433920504?

**Comment by carlescufi** [comment]
@zerotri @nashif merging this since it does have a relevant fix, we can continue the conversation in https://github.com/zephyrproject-rtos/zephyr/issues/10413#issuecomment-433920504

**Pull request merged**